### PR TITLE
HttpClient refactoring

### PR DESCRIPTION
**Pull request by merk**
- Adds the ability to the Logger to pass context through the exception() method
- Adds PSR HttpClientInterface to the HttpClient, allowing PSR7 Request -> Response (instead of a proxy method, using this interface seemed like a good idea..)
- Extracted out Logging, Exception Handling and StreamParsing into their own classes. The primary driver here was complexity > 15.

### CODE REVIEW

**Comment by merk** [comment]
I dont think the ServiceProvider is the best approach here, given most services that make Http calls will want to have a customised base url or headers to be used.

Maybe we implement some kind of ClientFactory?

File: src/HttpClient/ClientFactory.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+     * @param string|null $baseUri
+     * @param mixed[]|null $options
+     * @param \EoneoPay\Externals\Logger\Interfaces\LoggerInterface|null $logger
+     *
+     * @return \EoneoPay\Externals\HttpClient\Interfaces\ClientInterface
+     */
+    public function create(
+        ?string $baseUri = null,
+        ?array $options = null,
+        ?LoggerInterface $logger = null
```
Comment (merk) [inline src/HttpClient/ClientFactory.php]: I wonder if it should be passed in through the Factories constructor?

File: tests/HttpClient/ExceptionHandlerTest.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    /**
+     * Creates an instance
+     *
+     * @return \EoneoPay\Externals\HttpClient\ExceptionHandler
+     */
+    private function getInstance(): ExceptionHandler
+    {
+        return new ExceptionHandler();
```
Comment (edwardmurrell-loyaltycorp) [inline tests/HttpClient/ExceptionHandlerTest.php]: Does it make sense to have a `getInstance()` call when this takes literally not arguments?

File: tests/HttpClient/LoggingClientTest.php [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+                new Response(500, [], 'error')
+            )
+        ]);
+        $logger = new LogHandlerStub();
+
+        $instance = $this->createInstance($handler, $logger);
+
+        $previous = null;
+        try {
+            $instance->request('get', '/');
```
Comment (edwardmurrell-loyaltycorp) [inline tests/HttpClient/LoggingClientTest.php]: minor nitpick: lowercase get?

File: src/HttpClient/Client.php [hunk 4]
```diff
@@ -46,189 +55,84 @@ public function __construct(?Guzzle $client = null, ?LoggerInterface $logger = n
...
-            );
-        } catch (RequestException $exception) {
-            $response = $this->handleRequestException($exception);
-        } catch (GuzzleException $exception) {
-            // Covers any other guzzle exception
-            $response = new Response(['content' => $exception->getMessage()], 500);
+        $request = new Request(
+            $method,
+            $uri,
+            $options['headers'] ?? [],
```
Comment (edwardmurrell-loyaltycorp) [inline src/HttpClient/Client.php]: How come not having version as a method arg?

File: src/HttpClient/Exceptions/NetworkException.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        return static::DEFAULT_ERROR_CODE_RUNTIME;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getErrorSubCode(): int
+    {
+        return 0;
```
Comment (edwardmurrell-loyaltycorp) [inline src/HttpClient/Exceptions/NetworkException.php]: 0??????

File: src/HttpClient/Response.php [hunk 6]
```diff
@@ -4,20 +4,100 @@
...
-     * @var string
+     * Map of standard HTTP status code/reason phrases
+     *
+     * This array was copy and pasted shamelessly from the Zend Diactoros
+     * implementation because it wasnt part of the trait.
+     *
+     * @var string[]
      */
-    private $content;
+    private static $phrases = [
```
Comment (edwardmurrell-loyaltycorp) [inline src/HttpClient/Response.php]: I'm really surprised we can't already reference this from elsewhere in the codebase.

File: tests/HttpClient/ExceptionHandlerTest.php (hunk 2 above)
Comment (merk) [reply tests/HttpClient/ExceptionHandlerTest.php]: I dont know - seems following the same pattern is better but I dont really mind either way.

File: src/HttpClient/Client.php (hunk 4 above)
Comment (merk) [reply src/HttpClient/Client.php]: It comes from Guzzle, but I dont think Guzzle supports 2.0?

File: src/HttpClient/Response.php (hunk 6 above)
Comment (merk) [reply src/HttpClient/Response.php]: Yeah, its unfortunate but theres not much to be done if we want to conform to the Psr interfaces

Comment (edwardmurrell-loyaltycorp) [reply src/HttpClient/Response.php]: Ah right, fair enough.

File: src/HttpClient/Exceptions/NetworkException.php (hunk 5 above)
Comment (merk) [reply src/HttpClient/Exceptions/NetworkException.php]: 0 is a valid int ;)

File: tests/HttpClient/ExceptionHandlerTest.php (hunk 2 above)
Comment (edwardmurrell-loyaltycorp) [reply tests/HttpClient/ExceptionHandlerTest.php]: I'll accept it, as long as you rename the `$instance` vars in the tests that call this to be more meaningful.

**Pull request merged**